src/lem/coroutine/toplevel.rs

```diff
@@ -0,0 +1,339 @@
+            cs,
+            func,
+            not_dummy,
+            g,
+            store,
+            scope,
+            bound_allocations,
+            &mut next_acc,
+            &mut sub_provenances,
+        )
+        .unwrap();
+        let value = to_allocated_improper_list(cs, &res, g, store)?;
+        let provenance = self.synthesize_provenance(
+            ns!(cs, "provenance"),
+            g,
+            store,
+            value.clone(),
+            sub_provenances,
+            allocated_key,
+        )?;
+        Ok(((value, provenance), next_acc))
+    }
+
+    fn symbol(&self) -> Symbol {
+        self.name.clone()
+    }
+
+    fn from_ptr<CS: ConstraintSystem<F>>(cs: &mut CS, s: &Store<F>, ptr: &Ptr) -> Option<Self> {
+        ToplevelQuery::from_ptr(s, ptr).map(|q| q.to_circuit(cs, s))
+    }
+
+    fn dummy_from_index<CS: ConstraintSystem<F>>(
+        _cs: &mut CS,
+        _s: &Store<F>,
+        _index: usize,
+    ) -> Self {
+        unimplemented!()
+    }
+
+    #[allow(unused_variables)]
+    fn synthesize_args<CS: ConstraintSystem<F>>(
+        &self,
+        cs: &mut CS,
+        g: &GlobalAllocator<F>,
+        store: &Store<F>,
+    ) -> Result<AllocatedPtr<F>, SynthesisError> {
+        to_allocated_improper_list(cs, &self.args, g, store)
+    }
+}
+
+pub(crate) fn to_allocated_improper_list<F: LurkField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    args: &[AllocatedPtr<F>],
+    g: &GlobalAllocator<F>,
+    s: &Store<F>,
+) -> Result<AllocatedPtr<F>, SynthesisError> {
+    if args.is_empty() {
+        let nil = s.intern_nil();
+        let allocated_nil = g.alloc_ptr(cs, &nil, s);
+        Ok(allocated_nil)
+    } else {
+        // Iterator from last to first. Enumerate indices used for namespaces
+        let mut iter = args.iter().rev();
+        let mut args = iter.next().unwrap().clone();
+        for (i, arg) in iter.enumerate() {
+            let cs = ns!(cs, format!("arg:{i}"));
+            args = construct_cons(cs, g, s, arg, &args)?;
+        }
+        Ok(args)
+    }
+}
+
+pub(crate) fn to_improper_list<F: LurkField>(args: &[Ptr], s: &Store<F>) -> Ptr {
+    if args.is_empty() {
+        s.intern_nil()
+    } else {
+        // Iterator from last to first
+        let mut iter = args.iter().rev();
+        let mut args = *iter.next().unwrap();
+        for arg in iter {
+            args = s.cons(*arg, args);
+        }
+        args
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{func, state::user_sym};
+
+    use super::*;
+    use halo2curves::bn256::Fr as F;
+
+    #[test]
+    fn lem_coroutine_eval_test() {
+        let s = Arc::new(Store::<F>::default());
+        let factorial = func!(factorial(n): 1 => {
+            let zero = Num(0);
+            let one = Num(1);
+            let n_is_zero = eq_val(n, zero);
+            if n_is_zero {
+                return (one)
+            }
+            let m = sub(n, one);
+            let p = QUERY(factorial, m);
```

Comment:
To offload this detail that I saw on the call: `QUERY(factorial, m)` should probably be `QUERY("factorial", m)` so it doesn't look like we're passing a LEM variable as argument